### PR TITLE
Add Markdown export

### LINKED ISSUE #4243: Export to Markdown
Wallabag can export to various formats like PDF, EPub and others. Could be added a new option to export in markdown format? For taking notes and create brief documents, MD is usually preferred than ODT and HTML. It's a light format, there are a ton of apps compatible with it, and you can edit them in anything that can edit plain text.

### PR DESCRIPTION
**Pull request by nicosomb**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    |no
| Deprecations? | no
| Tests pass?   | yes
| Documentation |no
| Translation   | no
| CHANGELOG.md  | no
| License       | MIT

Should fix #4243

### CODE REVIEW

File: src/Helper/EntriesExport.php [hunk 1]
```diff
@@ -432,6 +433,34 @@ private function produceTxt()
...
     }
 
+    /**
+     * Dump a Markdown file.
+     *
+     * @return Response
+     */
+    private function produceMd()
+    {
+        $user = $this->tokenStorage->getToken() ? $this->tokenStorage->getToken()->getUser() : null;
```
Comment (yguedidi) [inline src/Helper/EntriesExport.php:443]: This seems unused after

File: src/Helper/EntriesExport.php [hunk 2]
```diff
@@ -432,6 +433,34 @@ private function produceTxt()
...
+        \assert($user instanceof User);
+
+        $content = '';
+        $converter = new HtmlConverter();
+        $converter->getConfig()->setOption('strip_tags', true);
+        foreach ($this->entries as $entry) {
```
Comment (yguedidi) [inline src/Helper/EntriesExport.php:449]: Does it make sense to support multiple entries for markdown export? I guess users will expect multiple md files, not all in a single one

File: templates/Entry/entries.html.twig [hunk 3]
```diff
@@ -105,6 +105,7 @@
...
             {% if craue_setting('export_txt') %}<li class="bold"><a class="waves-effect" href="{{ path('export_entries', {'category': current_route, 'format': 'txt', 'tag': current_tag, 'search_entry[term]': export_search_term, 'currentRoute': previous_route, 'entry': entry}) }}">TXT</a></li>{% endif %}
             {% if craue_setting('export_xml') %}<li class="bold"><a class="waves-effect" href="{{ path('export_entries', {'category': current_route, 'format': 'xml', 'tag': current_tag, 'search_entry[term]': export_search_term, 'currentRoute': previous_route, 'entry': entry}) }}">XML</a></li>{% endif %}
+            {% if craue_setting('export_md') %}<li class="bold"><a class="waves-effect" href="{{ path('export_entries', {'category': current_route, 'format': 'md', 'tag': current_tag, 'search_entry[term]': export_search_term, 'currentRoute': previous_route, 'entry': entry}) }}">MD</a></li>{% endif %}
```
Comment (yguedidi) [inline templates/Entry/entries.html.twig:108]: Is MD clear enough to end users?

File: tests/Controller/ExportControllerTest.php [hunk 4]
```diff
@@ -307,6 +307,23 @@ public function testXmlExport()
         $this->assertNotEmpty('updated_at', (string) $content->entry[0]->updated_at);
     }
 
+    public function testMdExport()
```
Comment (yguedidi) [inline tests/Controller/ExportControllerTest.php:310]: I'd add an assertion on the content

**Review by yguedidi — CHANGES_REQUESTED** [review]

**Comment by yguedidi** [comment]
Reviewed from phone 😅

File: src/Helper/EntriesExport.php (hunk 2 above)
Comment (nicosomb) [reply src/Helper/EntriesExport.php:449]: Why not, but we need to rewrite the Export system. That's how we handle other exports.

File: tests/Controller/ExportControllerTest.php [hunk 5]
```diff
@@ -307,6 +307,25 @@ public function testXmlExport()
...
+    {
+        $this->logInAs('admin');
+        $client = $this->getTestClient();
+
+        ob_start();
+        $crawler = $client->request('GET', '/export/all.md');
+        ob_end_clean();
```
Comment (yguedidi) [inline tests/Controller/ExportControllerTest.php:317]: I don't get why ob_* functions and a $crawler variable are needed here, I think they can be removed, just keep the call to request

**Review by yguedidi — CHANGES_REQUESTED** [review]

File: tests/Controller/ExportControllerTest.php [hunk 6]
```diff
@@ -307,6 +307,23 @@ public function testXmlExport()
         $this->assertNotEmpty('updated_at', (string) $content->entry[0]->updated_at);
     }
 
+    public function testMdExport()
+    {
+        $this->logInAs('admin');
+        $client = $this->getTestClient();
+
+        $crawler = $client->request('GET', '/export/all.md');
```
Comment (yguedidi) [inline tests/Controller/ExportControllerTest.php:315]: I would not store the crawler in a variable as it's unused

**Review by yguedidi — DISMISSED** [review]
One last comment, approved anyway as not critical so you can merge after the change

**Review by yguedidi — DISMISSED** [review]

File: migrations/Version20241112193044.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+declare(strict_types=1);
+
+namespace Application\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Wallabag\Doctrine\WallabagMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
```
Comment (j0k3r) [inline migrations/Version20241112193044.php:11]: Can you instead explain what it does?

File: src/Helper/EntriesExport.php [hunk 8]
```diff
@@ -432,6 +433,31 @@ private function produceTxt()
         );
     }
 
+    /**
+     * Dump a Markdown file.
+     *
+     * @return Response
+     */
+    private function produceMd()
```
Comment (j0k3r) [inline src/Helper/EntriesExport.php:441]: Can't we apply typing for new method?
```suggestion
     */
    private function produceMd(): Response
```

**Review by j0k3r — APPROVED** [review]

**Pull request merged**